File: python/adapter.py
"""Adapter docstring
"""
import sys
import argparse
import logging
import os
from clearblade.ClearBladeCore import System, Query
from clearblade.ClearBladeCore import cbLogs

#TODO Change the name of the adapter to a name associated with your implementation
ADAPTER_NAME = "MyAdapter"
CB_CONFIG = {}


def parse_env_variables(env):
    """Parse environment variables"""
    possible_vars = ["CB_SYSTEM_KEY", "CB_SYSTEM_SECRET", "CB_EDGE_NAME", "CB_PLATFORM_IP", "CB_EDGE_IP", "CB_ADAPTERS_ROOT_DIR", "CB_SERVICE_ACCOUNT", "CB_SERVICE_ACCOUNT_TOKEN"]
    
    for var in possible_vars:
        if var in env:
            print("Setting config from environment variable: " + var)
            CB_CONFIG[var] = env[var]

    #TODO Add implementation specific environment variables here


def parse_args(argv):
    """Parse the command line arguments"""

    parser = argparse.ArgumentParser(description='ClearBlade Adapter')
    parser.add_argument('-systemKey', dest="CB_SYSTEM_KEY", help='The System Key of the ClearBlade \
                        Plaform "System" the adapter will connect to.')

    parser.add_argument('-systemSecret', dest="CB_SYSTEM_SECRET", help='The System Secret of the \
                        ClearBlade Plaform "System" the adapter will connect to.')

    parser.add_argument('-deviceID', dest="deviceID", help='The id/name of the device that will be used for device \
                        authentication against the ClearBlade Platform or Edge, defined \
                        within the devices table of the ClearBlade platform.')

    parser.add_argument('-activeKey', dest="activeKey", help='The active key of the device that will be used for device \
                        authentication against the ClearBlade Platform or Edge, defined within \
                        the devices table of the ClearBlade platform.')

    parser.add_argument('-cb_service_account', dest="CB_SERVICE_ACCOUNT", help='The id/name of the device service accountthat will be used for \
                        authentication against the ClearBlade Platform or Edge, defined \
                        within the devices table of the ClearBlade platform.')

    parser.add_argument('-cb_service_account_token', dest="CB_SERVICE_ACCOUNT_TOKEN", help='The token of the device service account that will be used for device \
                        authentication against the ClearBlade Platform or Edge, defined within \
                        the devices table of the ClearBlade platform.')

    parser.add_argument('-httpUrl', dest="httpURL", default="http://localhost", \
                        help='The HTTP URL of the ClearBlade Platform or Edge the adapter will \
                        connect to. The default is https://localhost.')

    parser.add_argument('-httpPort', dest="httpPort", default="9000", \
                        help='The HTTP Port of the ClearBlade Platform or Edge the adapter will \
                        connect to. The default is 9000.')

    parser.add_argument('-messagingUrl', dest="messagingURL", default="localhost", \
                        help='The MQTT URL of the ClearBlade Platform or Edge the adapter will \
                        connect to. The default is https://localhost.')

    parser.add_argument('-messagingPort', dest="messagingPort", type=int, default=1883, \
                        help='The MQTT Port of the ClearBlade Platform or Edge the adapter will \
                        connect to. The default is 9000.')

    parser.add_argument('-adapterSettingsCollection', dest="adapterSettingsCollectionName", \
                        default="", \
                        help='The name of the ClearBlade Platform data collection which contains \
                        runtime configuration settings for the adapter. The default is "".')

    parser.add_argument('-adapterSettingsItem', dest="adapterSettingsItemID", default="", \
                        help='The "item_id" of the row, within the ClearBlade Platform data \
                        collection which contains runtime configuration settings, that should \
                        be used to configure the adapter. The default is "".')

    parser.add_argument('-topicRoot', dest="adapterTopicRoot", default="", \
                        help='The root of MQTT topics this adapter will subscribe and publish to. \
                        The default is "".')

    parser.add_argument('-logLevel', dest="logLevel", default="INFO", choices=['CRITICAL', \
                        'ERROR', 'WARNING', 'INFO', 'DEBUG'], help='The level of logging that \
                        should be utilized by the adapter. The default is "INFO".')

    parser.add_argument('-logCB', dest="logCB", default=False, action='store_true',\
                        help='Flag presence indicates logging information should be printed for \
                        ClearBlade libraries.')

    parser.add_argument('-logMQTT', dest="logMQTT", default=False, action='store_true',\
                        help='Flag presence indicates MQTT logs should be printed.')

    #TODO Add implementation specific command line arguments here

    args = vars(parser.parse_args(args=argv[1:]))
    for var in args:
        if args[var] != "" and args[var] != None:
            print("Setting config from command line argument: " + var)
            CB_CONFIG[var] = args[var]


def check_required_config():
    """Verify all required config options were provided via environment variables or command line arguments"""
    if "CB_SYSTEM_KEY" not in CB_CONFIG:
        logging.error("System Key is required, can be provided with CB_SYSTEM_KEY environment variable or --systemKey command line argument")
        exit(-1)
    if not "CB_SYSTEM_SECRET" in CB_CONFIG:
        logging.error("System Secret is required, can be provided with CB_SYSTEM_SECRET environment variable or --systemSecret command line argument")
        exit(-1)

    if "deviceID" in CB_CONFIG and CB_CONFIG["deviceID"] != "" and CB_CONFIG["deviceID"] != None:
        if "activeKey" not in CB_CONFIG:
            logging.error("Device Active Key is required when a deviceID is specified, can be provided with the --activeKey command line argument")
            exit(-1)
    elif "CB_SERVICE_ACCOUNT" in CB_CONFIG and CB_CONFIG["CB_SERVICE_ACCOUNT"] != "" and CB_CONFIG["CB_SERVICE_ACCOUNT"] != None:
        if "CB_SERVICE_ACCOUNT_TOKEN" not in CB_CONFIG:
            logging.error("Device Service Account Token is required when a Service Account is specified, can be provided with the CB_SERVICE_ACCOUNT_TOKEN enviornment variable or --cb_service_account_token command line argument")
            exit(-1)
    else:
        logging.error("Device ID/Active Key or Service Account Name and Token are required")
        exit(-1)
    logging.debug("Adapter Config Looks Good!")
    
    

def setup_custom_logger(name):
    """Create a custom logger"""
    #logging.debug("Begin setup_custom_logger")

    formatter = logging.Formatter(fmt='%(asctime)s %(levelname)-8s %(message)s', \
                                  datefmt='%m-%d-%Y %H:%M:%S %p')
    handler = logging.StreamHandler(stream=sys.stdout)
    handler.setFormatter(formatter)
    logger = logging.getLogger(name)
    logging.basicConfig(level=os.environ.get("LOGLEVEL", CB_CONFIG['logLevel']))
    logger.addHandler(handler)

    #logging.debug("End setup_custom_logger")

    return logger


def get_adapter_config():
    """Retrieve the runtime configuration for the adapter from a ClearBlade Platform data \
    collection"""
    logging.debug("Begin get_adapter_config")

    logging.debug('Retrieving the adapter configuration from data collection %s', \
        CB_CONFIG['adapterSettingsCollectionName'])

    collection = CB_SYSTEM.Collection(CB_AUTH,
                                      collectionName=CB_CONFIG['adapterSettingsCollectionName'])

    the_query = Query()
    if CB_CONFIG['adapterSettingsItemID'] != "":
        the_query.equalTo("item_id", CB_CONFIG['adapterSettingsItemID'])

    rows = collection.getItems(the_query)

    # Iterate through rows and display them
    for row in rows:
        #TODO - Implement implementation specific logic
        logging.debug(row)

    logging.debug("End get_adapter_config")


#########################
#BEGIN MQTT CALLBACKS
#########################

#TODO - Comment out or remove the MQTT CALLBACKS block if your adapter will not be utilizing MQTT
#TODO - Comment out or remove any of the following MQTT callbacks your adapter implementation does not need

def on_connect(mqtt_client, userdata, flags, result_code):
    """MQTT callback invoked when a connection is established with a broker"""
    logging.debug("Begin on_connect")
    logging.info("Connected to ClearBlade Platform MQTT broker")

    #When the connection to the broker is complete, set up any subscriptions that are needed

    #TODO - If your implementation needs to subscribe to topics, uncomment this block and add your logic here
    # log.Printf("[DEBUG] Begin Configuring Subscription(s)")
    # client.subscribe(CB_CONFIG['adapterTopicRoot'+ '/mySubTopic'])
    global EXIT_APP
    EXIT_APP = True

    logging.debug("End on_connect")

def on_disconnect(mqtt_client, userdata, result_code):
    """MQTT callback invoked when a connection to a broker is lost"""
    logging.debug("Begin on_disconnect")

    if result_code != 0:
        logging.warning("Connection to CB Platform MQTT broker was lost, result code = %s", \
                        result_code)

    #We don't need to worry about manally re-initializing the mqtt client. The auto reconnect
    # logic will automatically try and reconnect. The reconnect interval could be as much as
    # 20 minutes.

    #TODO - Add implementation specific logic here

    logging.debug("End on_disconnect")

def on_subscribe(mqtt_client, userdata, mid, granted_qos):
    """MQTT callback invoked when a subscription has been made"""
    logging.debug("Begin on_subscribe")

    #TODO Add implementation specific logic here 


    logging.debug("End on_subscribe")

def on_unsubscribe(mqtt_client, userdata, mid):
    """MQTT callback invoked when a subscription has been made"""
    logging.debug("Begin on_unsubscribe")

    #TODO Add implementation specific logic here 


    logging.debug("End on_unsubscribe")

def on_publish(mqtt_client, userdata, mid):
    """MQTT callback invoked when a message  has been published"""
    logging.debug("Begin on_publish")

    #TODO Add implementation specific logic here 


    logging.debug("End on_publish")


def on_message(mqtt_client, userdata, mid):
    """MQTT callback invoked when a message is received"""
    logging.debug("Begin on_message")

    #TODO Add implementation specific logic here 


    logging.debug("End on_message")


def on_log(mqtt_client, userdata, level, buf):
    """MQTT callback invoked when MQTT prints a logging statement"""
    logging.debug("Begin on_log")

    logging.debug("End on_log")

#########################
#END MQTT CALLBACKS
#########################


#Main Loop
if __name__ == '__main__':
    #logging.info("Validating command line arguments")
    global EXIT_APP
    EXIT_APP = False

    parse_env_variables(os.environ)
    parse_args(sys.argv)

    LOGGER = setup_custom_logger(ADAPTER_NAME)
    
    check_required_config()

    logging.info("Intializing ClearBlade device client")
    logging.debug("System Key = %s", CB_CONFIG['CB_SYSTEM_KEY'])
    logging.debug("System Secret = %s", CB_CONFIG['CB_SYSTEM_SECRET'])
    logging.debug("HTTP URL = %s", CB_CONFIG['httpURL'] + ":" + CB_CONFIG['httpPort'])

    CB_SYSTEM = System(CB_CONFIG['CB_SYSTEM_KEY'], CB_CONFIG['CB_SYSTEM_SECRET'], CB_CONFIG['httpURL'] + \
                       ":" + CB_CONFIG['httpPort'])



    logging.info("Authenticating to ClearBlade")

    CB_AUTH = None

    if "deviceID" in CB_CONFIG:
        logging.info("Authenticating to ClearBlade with Device ID and Active Key")
        CB_AUTH = CB_SYSTEM.Device(CB_CONFIG['deviceID'], CB_CONFIG['activeKey'])
    elif "CB_SERVICE_ACCOUNT" in CB_CONFIG:
        logging.info("Authenticating to ClearBlade with Device Service Account and Token")
        CB_AUTH = CB_SYSTEM.Device(CB_CONFIG["CB_SERVICE_ACCOUNT"], authToken=CB_CONFIG["CB_SERVICE_ACCOUNT_TOKEN"])
    else:
        logging.error("No device info provided to authenticate to ClearBlade as")
        exit(-1)


    #Retrieve the adapter configuration
    if "adapterSettingsCollectionName" in CB_CONFIG and CB_CONFIG['adapterSettingsCollectionName'] != "":
        logging.info("Retrieving the adapter configuration settings")
        get_adapter_config()

    #########################
    #BEGIN MQTT SPECIFIC CODE
    #########################

    #TODO - Comment out or remove the MQTT SPECIFIC CODE block if your adapter will not be utilizing MQTT

    #Connect to the message broker
    logging.info("Initializing the ClearBlade message broker")
    CB_MQTT = CB_SYSTEM.Messaging(CB_AUTH)

    #TODO - Comment out or remove any of the following 7 lines that you do not need 
    CB_MQTT.on_log = on_log
    CB_MQTT.on_connect = on_connect
    CB_MQTT.on_disconnect = on_disconnect
    CB_MQTT.on_subscribe = on_subscribe
    CB_MQTT.on_unsubscribe = on_unsubscribe
    CB_MQTT.on_publish = on_publish
    CB_MQTT.on_message = on_message

    logging.info("Connecting to the ClearBlade message broker")
    CB_MQTT.connect() #Connect to the msg broker

    #END MQTT SPECIFIC CODE

    while not EXIT_APP:
        try:
            #TODO - Add implementation specific logic here
            pass
        except KeyboardInterrupt:
            EXIT_APP = True
            CB_MQTT.disconnect()
            sys.exit(0)
        except Exception as e:
            logging.info ("EXCEPTION:: %s", str(e))
        #finally:
            #TODO - Add implementation specific logic here

File: golang/adapter.go
package main

import (
	"flag"
	"log"
	"os"
	"strings"
	"time"

	lumberjack "gopkg.in/natefinch/lumberjack.v2"

	cb "github.com/clearblade/Go-SDK"
	MQTT "github.com/eclipse/paho.mqtt.golang"
	"github.com/hashicorp/logutils"
)

var (
	httpURL         string //Variable to hold the HTTP URL of the ClearBlade Platform or Edge the adapter will connect to
	httpPort        string //Variable to hold the HTTP Port of the ClearBlade Platform or Edge the adapter will connect to
	messagingURL    string //Variable to hold the MQTT URL of the ClearBlade Platform or Edge the adapter will connect to
	messagingPort   string //Variable to hold the MQTT Port of the ClearBlade Platform or Edge the adapter will connect to
	sysKey          string //Variable to hold the System Key of the ClearBlade Plaform "System" the adapter will connect to
	sysSec          string //Variable to hold the System Secret of the ClearBlade Plaform "System" the adapter will connect to
	deviceID        string //Variable to hold the id/name of the device that will be used for device authentication against the ClearBlade Platform or Edge
	deviceActiveKey string //Variable to hold the active key of the device that will be used for device authentication against the ClearBlade Platform or Edge

	adapterSettingsCollectionName string //Variable to hold the name of the ClearBlade Platform data collection which contains runtime configuration settings for the adapter
	adapterSettingsItemID         string //Variable to hold the "item_id" of the row, within the ClearBlade Platform data collection which contains runtime configuration settings, that should be used to configure the adapter

	adapterTopicRoot string //Variable to hold the root of MQTT topics this adapter will subscribe and publish to

	deviceProvisionSvc    string //Variable to hold the name of a service that can be invoked to provision IoT devices within the ClearBlade Platform or Edge
	deviceHealthSvc       string //Variable to hold the name of a service that can be invoked to provide the health of an IoT device to the ClearBlade Platform or Edge
	deviceLogsSvc         string //Variable to hold the name of a service that can be invoked to provide IoT device logging information to the ClearBlade Platform or Edge
	deviceStatusSvc       string //Variable to hold the name of a service that can be invoked to provide the status of an IoT device to the ClearBlade Platform or Edge
	deviceDecommissionSvc string //Variable to hold the name of a service that can be invoked to decommission IoT devices within the ClearBlade Platform or Edge

	logLevel string
	logMQTT  bool

	deviceClient *cb.DeviceClient
)

const (
	platURL  = "http://localhost"
	platPort = "9000"
	messURL  = "localhost"
	messPort = "1883"
)

func init() {
	flag.StringVar(&sysKey, "systemKey", "", "system key (required)")
	flag.StringVar(&sysSec, "systemSecret", "", "system secret (required)")
	flag.StringVar(&deviceID, "deviceID", "", "id/name of device used for device authentication (required)")
	flag.StringVar(&deviceActiveKey, "deviceActiveKey", "", "active key of the device used for device authentication (required)")

	flag.StringVar(&httpURL, "httpURL", platURL, "ClearBlade Platform or ClearBlade Edge HTTP URL. Will default to \"http://localhost\". (optional)")
	flag.StringVar(&httpPort, "httpPort", platPort, "ClearBlade Platform or ClearBlade Edge http port. Will default to \"9000\". (optional)")

	flag.StringVar(&messagingURL, "messagingURL", messURL, "ClearBlade Platform or ClearBlade Edge messaging URL. Will default to \"localhost\". (optional)")
	flag.StringVar(&messagingPort, "messagingPort", messPort, "ClearBlade Platform or ClearBlade Edge messaging URL. Will default to \"1883\". (optional)")

	flag.StringVar(&adapterSettingsCollectionName, "adapterSettingsCollection", "", "The name of a data collection containing runtime settings for the adapter (optional)")
	flag.StringVar(&adapterSettingsItemID, "adapterSettingsItem", "", "The \"item_id\" of a specific row within the adapter runtime settings data collection to retrieve (optional)")

	flag.StringVar(&adapterTopicRoot, "topicRoot", "", "The root MQTT topic the adapter should use when publishing or subscribing. (optional)")

	flag.StringVar(&deviceProvisionSvc, "deviceProvisionSvc", "", "The name of a ClearBlade Platform code service that will receive device provisioning requests. (optional)")
	flag.StringVar(&deviceHealthSvc, "deviceHealthSvc", "", "The name of a ClearBlade Platform code service that will receive device health requests. (optional)")
	flag.StringVar(&deviceLogsSvc, "deviceLogsSvc", "", "The name of a ClearBlade Platform code service that will receive device log requests. (optional)")
	flag.StringVar(&deviceStatusSvc, "deviceStatusSvc", "", "The name of a ClearBlade Platform code service that will receive device status requests. (optional)")
	flag.StringVar(&deviceDecommissionSvc, "deviceDecommissionSvc", "", "The name of a ClearBlade Platform code service that will receive device decommissioning requests. (optional)")

	flag.StringVar(&logLevel, "logLevel", "warn", "The level of logging to use. Available levels are 'debug', 'warn', 'error' (optional)")
	flag.BoolVar(&logMQTT, "logMQTT", false, "Indicates MQTT log information should be written")

	//TODO - Add code to handle any other command line arguments specific to your implmentation
}

func usage() {
	log.Printf("Usage: adapter [options]\n\n")
	flag.PrintDefaults()
}

func validateFlags() {
	log.Printf("Validating command line options")

	flag.Parse()

	//TODO If any implmentation specific REQUIRED command line parameters have been added,
	//add the associated variables to this if statement
	if sysKey == "" ||
		sysSec == "" ||
		deviceID == "" ||
		deviceActiveKey == "" {

		log.Printf("[ERROR] Missing required flags\n\n")
		flag.Usage()
		os.Exit(1)
	}

	if logLevel != "error" && logLevel != "warn" && logLevel != "debug" {
		log.Printf("[ERROR] Invalid log level specified\n\n")
		flag.Usage()
		os.Exit(1)
	}

	//TODO Add any other implementation specific command line argument validation logic here
}

//ClearBlade Device Client init helper
func initCbDeviceClient() {
	log.Printf("Initializing ClearBlade device client")

	log.Printf("[DEBUG] Setting System Key to %s", sysKey)
	log.Printf("[DEBUG] Setting System Secret to %s", sysSec)
	log.Printf("[DEBUG] Setting device ID to %s", deviceID)
	log.Printf("[DEBUG] Setting active key to %s", deviceActiveKey)
	log.Printf("[DEBUG] Setting platform URL to %s", httpURL+":"+httpPort)
	log.Printf("[DEBUG] Setting messaging URL to %s", messagingURL+":"+messagingPort)

	deviceClient = cb.NewDeviceClientWithAddrs(httpURL+":"+httpPort, messagingURL+":"+messagingPort, sysKey, sysSec, deviceID, deviceActiveKey)

	for err := deviceClient.Authenticate(); err != nil; {
		log.Printf("[WARN] Error authenticating to platform: %s", err.Error())

		//TODO Add logic to differentiate between timeout errors and bad username/password errors
		//Only retry authentication if a timeout or connection issue occurs

		log.Printf("[WARN] Will retry in 1 minute...")

		// sleep 1 minute
		time.Sleep(time.Duration(time.Minute * 1))
		err = deviceClient.Authenticate()
	}

	log.Printf("[DEBUG] Initializing MQTT with callbacks")
	var callbacks = &cb.Callbacks{OnConnectionLostCallback: OnConnectLost, OnConnectCallback: OnConnect}

	//TODO - replace "myAdapter" with the name of your adapter
	if mqtterr := deviceClient.InitializeMQTTWithCallback("myAdapter"+deviceID, "", 30, nil, nil, callbacks); mqtterr != nil {
		log.Fatalf("[ERROR] initCbClient: Unable to initialize MQTT connection to broker %s: %s",
			messagingURL+":"+messagingPort, mqtterr.Error())
	}

	log.Printf("Finished initializing ClearBlade device client")
}

//getAdapterConfig - Retrieve adapter runtime configuration parameters from a platform/edge data collection
func getAdapterConfig() error {
	log.Printf("Retrieving adapter runtime settings")

	query := &cb.Query{}

	if adapterSettingsItemID != "" {
		log.Printf("[DEBUG] Adapter settings item id = %s", adapterSettingsItemID)
		query := cb.NewQuery()
		query.EqualTo("item_id", adapterSettingsItemID)
	}

	//Retrieve the adapter configuration row. Passing a nil query results in all rows being returned
	results, err := deviceClient.GetDataByName(adapterSettingsCollectionName, query)
	if err != nil {
		log.Printf("[WARN] Adapter configuration could not be retrieved. Error: %s", err.Error())
		return err
	}

	log.Printf("[DEBUG] Adapter settings retrieved: %#v", results)

	//TODO - Add implemenation specific logic
	//example: myVar = results["DATA"].([]interface{})[0].(map[string]interface{})["myCollectionColumn"].(string)

	log.Printf("Finished retrieving adapter runtime settings")

	return nil
}

//OnConnectLost - MQTT callback invoked when a connection to a broker is lost
func OnConnectLost(client MQTT.Client, connerr error) {
	log.Printf("[WARN] Connection to broker was lost: %s", connerr.Error())

	//We don't need to worry about manally re-initializing the mqtt client. The auto reconnect logic will
	//automatically try and reconnect. The reconnect interval could be as much as 20 minutes.

	//TODO - Add any implementation logic here
}

//OnConnect - MQTT callback invoked when a connection is established with a broker
//When the connection to the broker is complete, set up the subscriptions
func OnConnect(client MQTT.Client) {
	log.Printf("Connected to ClearBlade Platform MQTT broker")

	//TODO - If your implementation needs to subscribe to topics, uncomment this block and add your logic here
	// log.Printf("[DEBUG] Begin Configuring Subscription(s)")

	// for mySubscribeChannel, err = deviceClient.Subscribe(adapterTopicRoot+"/"+subscribeTopic, messagingQos); err != nil; {
	// 	log.Printf("[WARN] Error subscribing to topics: %s", err.Error())

	// 	//Wait 30 seconds and retry
	// 	log.Printf("[DEBUG] Waiting 30 seconds to retry subscriptions")
	// 	time.Sleep(time.Duration(30 * time.Second))
	// 	mySubscribeChannel, err = deviceClient.Subscribe(adapterTopicRoot+"/"+subscribeTopic, messagingQos)
	// }
}

func main() {
	flag.Usage = usage
	validateFlags()

	//////////////////////////////////////////
	//BEGIN Custom logfile implemenation block
	//////////////////////////////////////////

	// If you do not wish to implement rolling log files, comment out/remove this block of code

	//TODO - Rename the file "/var/log/adapter.log" to provide a name that describes the adapter
	logFileName := "/var/log/adapter.log"

	_, err := os.OpenFile(logFileName, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)

	if err != nil {
		log.Fatalf("error opening file: %s", err.Error())
	}

	log.SetFlags(log.LstdFlags | log.Lshortfile)

	//TODO - Modify these values to fit the needs of your adapter
	var lumberjackLogger = &lumberjack.Logger{
		Filename:   logFileName,
		MaxSize:    10, // megabytes
		MaxBackups: 5,
		MaxAge:     28, //days
	}

	filter := &logutils.LevelFilter{
		Levels:   []logutils.LogLevel{"DEBUG", "WARN", "ERROR"},
		MinLevel: logutils.LogLevel(strings.ToUpper(logLevel)),
		Writer:   lumberjackLogger,
	}
	log.Printf("Logging output is being redirected to %s. View %s for adapter log details.", logFileName, logFileName)
	log.SetOutput(filter)

	//////////////////////////////////////////
	//END Custom logfile implemenation block
	//////////////////////////////////////////

	///////////////////////////
	//BEGIN MQTT log file block
	///////////////////////////

	//Initialize MQTT Logging
	if logMQTT {
		//Always show critical log messages
		MQTT.CRITICAL = log.New(lumberjackLogger, "CRITICAL: ", log.Ldate|log.Ltime|log.Lshortfile)
		if logLevel == "debug" {
			MQTT.DEBUG = log.New(lumberjackLogger, "DEBUG: ", log.Ldate|log.Ltime|log.Lshortfile)
		}
		if logLevel == "warn" || logLevel == "debug" {
			MQTT.WARN = log.New(lumberjackLogger, "WARN: ", log.Ldate|log.Ltime|log.Lshortfile)
		}
		if logLevel == "error" || logLevel == "warn" || logLevel == "debug" {
			MQTT.ERROR = log.New(lumberjackLogger, "ERROR: ", log.Ldate|log.Ltime|log.Lshortfile)
		}
	}

	///////////////////////////
	//END MQTT log file block
	///////////////////////////

	//Initialize CB device client
	initCbDeviceClient()

	if adapterSettingsCollectionName != "" {
		//Retrieve adapter config info
		getAdapterConfig()
	}

	//TODO - Add your implemenation specific code here
	//At this point, the CB device client will have been
	//initialized and authenticated to the ClearBlade Platform or ClearBlade Edge
	select {}
}
